go-controller/pkg/ovn/pods.go

```diff
@@ -497,3 +620,150 @@ func (oc *Controller) getPortAddresses(nodeName, portName string) (net.HardwareA
 	}
 	return podMac, podIPNets, nil
 }
+
+func (oc *Controller) addPodExternalGW(pod *kapi.Pod) error {
+	type Network struct {
+		Name string
+		Ips  []string
+	}
+	routingNamespaceAnnotation := pod.Annotations[routingNamespaceAnnotation]
+	if routingNamespaceAnnotation == "" {
+		return nil
+	}
+	klog.Infof("External gateway pod: %s, detected for namespace(s) %s", pod.Name, routingNamespaceAnnotation)
+	var foundGws []net.IP
+	if routingNetworkAnnotation == "" && pod.Spec.HostNetwork {
+		for _, podIP := range pod.Status.PodIPs {
+			ip := net.ParseIP(podIP.IP)
+			if ip != nil {
+				foundGws = append(foundGws, ip)
+			}
+		}
+	} else if routingNetworkAnnotation != "" {
+		var multusNetworks []Network
+		err := json.Unmarshal([]byte(pod.ObjectMeta.Annotations["k8s.v1.cni.cncf.io/network-status"]), &multusNetworks)
+		if err != nil {
+			return fmt.Errorf("unable to unmarshall annotation k8s.v1.cni.cncf.io/network-status on pod %s: %v", pod.Name, err)
+		}
+		for _, multusNetwork := range multusNetworks {
+			if multusNetwork.Name == routingNetworkAnnotation {
+				for _, gwIP := range multusNetwork.Ips {
+					ip := net.ParseIP(gwIP)
+					if ip != nil {
+						foundGws = append(foundGws, ip)
+					}
+				}
+			}
+		}
+	} else {
+		klog.Errorf("Ignoring pod %s as an external gateway candidate. Invalid combination "+
+			"of host network: %t and routing-network annotation: %s", pod.Name, pod.Spec.HostNetwork, routingNetworkAnnotation)
+		return nil
+	}
+
+	// if we found any gateways then we need to update current pods routing in the relevant namespace
+	if len(foundGws) == 0 {
+		klog.Warningf("No valid gateway IPs found for requested external gateway pod: %s", pod.Name)
+		return nil
+	}
+
+	for _, namespace := range strings.Split(routingNamespaceAnnotation, ",") {
+		nsInfo, err := oc.waitForNamespaceLocked(namespace)
+		if err != nil {
+			return err
+		}
+		defer nsInfo.Unlock()
+		nsInfo.routingExternalPodGWs[pod.Name] = foundGws
+		existingPods, err := oc.watchFactory.GetPods(namespace)
+		if err != nil {
+			return fmt.Errorf("failed to get all the pods for namespace %s, error: %v", namespace, err)
+		}
+		for _, gwIP := range foundGws {
+			for _, pod := range existingPods {
+				for _, podIP := range pod.Status.PodIPs {
+					var mask string
+					if utilnet.IsIPv6(net.ParseIP(podIP.IP)) {
```

Comment:
If we're gonna keep using this, why not:

```
func getIPFullMask(ip net.IP) string {
    if utilnet.IsIPv6(net.ParseIP(podIP.IP)) {
        return IPv6FullMask
    }
    return IPv4FullMask
}
```

Reply:
done